src/foam/nanos/medusa/BlockingDAO.js

```diff
@@ -0,0 +1,170 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.medusa',
+  name: 'BlockingDAO',
+  extends: 'foam.dao.ProxyDAO',
+
+  abstract: true,
+  
+  documentation: `All DAO operations will blocked while condition is true, and unblocked when a particular CMD is recieved.`,
+
+  javaImports: [
+    'foam.dao.DAO',
+    'foam.nanos.pm.PM',
+    'foam.nanos.logger.Logger',
+    'foam.nanos.logger.Loggers',
+    'java.util.concurrent.atomic.AtomicLong'
+  ],
+
+  javaCode: `
+    private Object lock_ = new Object();
+    private AtomicLong blocked_ = new AtomicLong();
+  `,
+
+  constants: [
+    {
+      documentation: 'Return number of threads currently blocked',
+      name: 'BLOCKED_CMD',
+      type: 'String',
+      value: 'BLOCKED_CMD'
+    },
+    {
+      documentation: 'Notify all threads',
+      name: 'UNBLOCK_CMD',
+      type: 'String',
+      value: 'BLOCKED_CMD'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'find_',
+      javaCode: `
+      // Loggers.logger(x, this).debug("find");
+      block(x);
+      return getDelegate().find_(x, id);
+      `
+    },
+    {
+      name: 'select_',
+      javaCode: `
+      // Loggers.logger(x, this).debug("select");
+      block(x);
+      return getDelegate().select_(x, sink, skip, limit, order, predicate);
+      `
+    },
+    {
+      name: 'put_',
+      javaCode: `
+      // Loggers.logger(x, this).debug("put");
+      block(x);
+      return getDelegate().put_(x, obj);
+      `
+    },
+    {
+      name: 'remove_',
+      javaCode: `
+      // Loggers.logger(x, this).debug("remove");
+      block(x);
+      return getDelegate().remove_(x, obj);
+      `
+    },
+    {
+      name: 'removeAll_',
+      javaCode: `
+      // Loggers.logger(x, this).debug("removeAll");
+      block(x);
+      getDelegate().removeAll_(x, skip, limit, order, predicate);
+      `
+    },
+    {
+      name: 'cmd_',
+      javaCode: `
+      if ( obj != null ) {
+        if ( UNBLOCK_CMD.equals(obj) ) {
+           /* obj.equals(getUnblockCmd()) ||*/
+          Loggers.logger(x, this).info("cmd", "unblock", "received", "blocked", blocked_.get());
+          unblock(x);
+        }
+        if ( BLOCKED_CMD.equals(obj) ) {
+          return blocked_.get();
+        }
+      }
+      return getDelegate().cmd_(x, obj);
+      `
+    },
+    {
+      name: 'block',
+      args: [
```

Comment:
You could just say: args: 'Context x'

Reply:
No, I much prefer in verbose form for git deltas.